### ISSUE TITLE
Ideas to massively improve the queue system (& some new feature requests)

### ISSUE THREAD

**Issue by AdvaitaSprite** [easydiffusion/easydiffusion#511]
Absolutely love the project (you all are doing incredible work, and the fact that you make it freely available wins my undying respect), but just a couple of requests to improve upon it. I have a couple of other ideas I'll post about at a later date, but for the sake of coherency I'll focus this post around relatively simple changes to the queue system.

I think these basic changes to the queue would constitute the most game-changing improvements (on a sheer functionality/usability/quality of life level), short of adding entire new features, which I'll tack on at the end, but I suspect these alterations to the queue would be easier to implement:

1.  - It's discouraging & frustrating when you spend hours working on an extensive queue which you'll never be able to recreate if you lose it (with a wide variety of different modifier combinations, settings, etc.), only to either have the window crash or get closed for some reason [has only happened to me once upon slightly overloading the queue, but it still is tragic when you're in a flow & suddenly lose everything], so **making it possible to save a queue, close the window, and still come back to it later**, as well as making it so that you can **pause at will, and still be able to start the queue back where you left it** would be just a MASSIVE QoL improvement. 

2.  - It can also be frustrating when you create a bunch of batches with slight variations on a prompt (different modifiers/settings/changes to the body of the prompt)... Only to realize after the fact that you forgot some very basic modifier, like "beautiful lighting," or "colorful," or "intricate detail," which could sabotage the whole queue you just lined up. Or maybe you just immediately wish you'd asked for "an elaborate paper model of. . ." instead of "A realistic oil painting of. . .". 

      The amount of work it would take to write/copy down each item, erase them from queue, and then recreate the whole queue down to the settings, with the missing modifiers added in, would just be exhausting. It would be much easier if we could simply **edit the prompt/modifiers/settings for batches in the queue which have not yet begun rendering**. That way, you could just run down the list very quickly and add what you need to without disrupting the queue (as long as it hasn't already started to generate the image for that batch I don't see why it shouldn't be possible)!
      
3.  - I also had at least one experience where it seemed as if, when you try to add an inpainting batch to the queue, it cancels all non-inpainting batches already in the queue without warning, so making it so that you can either (a) seamlessly include both in-painting &  normal txt2img/img2img batches to the queue without issue, or (b) if it can't be helped, then at least add a warning upon switching to or from in-painting mode if doing so will erase your current queued items.

The other ideas for features I've been thinking about, which I'd love to see implemented in this UI, but which would constitute major additions & not something I necessarily expect any time soon:

- In my wildest dreams, a tab supporting stable-dreamfusion (never seen it implemented within a txt2img/img2img stable-diffusion UI, so I assume it would be a challenge, but it'd also be groundbreaking, so maybe integrating it could be someone's pet project, whether on a different experimental fork or in the beta mode or something).

- Adding support for textual inversion and aesthetic embeddings (this would be really big, as it would enable you to create a symbol representing a variable which SD would "understand" as referring to a specific person, aesthetic style, or whatever else, in your prompt. 

- The option to use CodeFormer [a considerably more powerful face-fixing algorithm than GFPGAN in my experience] instead of GFPGAN & RealESRGAN to fix faces/details & upscale, respectively [this would require you to be able to access the CodeFormer settings, which consist of just one variable and a couple of toggle boxes for "enhance background in addition to faces" and a place to type in the factor by which to upscale].

Those last two are already integrated into, for example, AUTOMATIC1111's webui, so you can look at how they're handled there for inspiration/guidance [and there are other features implemented there which would be cool to see eventually, such as hypernetwork support, but I haven't used that, so it's not a priority for me]. However, I love this UI; the queue system overall works very well (and with the above improvements would be pretty flawless). I love that you can install this UI locally and eliminate the need for any API calls or cloud GPUs... So I'd much rather see new features added & developed upon here rather than switch to any kind of web-based UI. 

I also love that you've managed to get the VRAM requirements down to where my 4gb 1050ti can make use of this powerful algorithm. So **if any of these features would jeopardize that-- increasing the minimum VRAM requirement above 4gb--** then I'd _strongly_ appreciate the UI **warning you about that before selecting them, and making them optional**, so that they don't suddenly make the UI unusable for me & many others.

Thanks so much. I know that's a significant amount for one post, but I've given this some thought & felt these possible improvements were worthy of mention. Anyone who works towards implementing them, I am tremendously grateful. This is my first time requesting features, so I hope my format is okay (including more than one idea here), & I hope the clarity & specificity was sufficient to justify the length.

Edit: Oh, the one other thing I'd REALLY appreciate would be the ability to choose how many steps you add when you use the current "25 more steps" button. 25 is actually a LOT of steps, and very often if you use good settings the first time, 25 more will drastically over-shoot your target (depending on what sampler you're using). Whether it's toggle, a dropdown box where you can choose among increments of 5 steps, or just a place to type in how many more steps you want to add to a given image, having options there would be huge. I love the ease of being able to just click "x more steps" when you want to refine an image you like, but if I use a reasonable setting the first time around [e.g. 40 or 45 steps with dpm2] then 25 additional steps is pretty much guaranteed to burn the bejeezus out of it. But when you're working with whole queues of different batches, it can be way too much work to do the only alternative, which is to choose "use as initial image," then go back to the queue settings and tell it a different number of steps in an img2img workflow. This may be simple for one image, but if there are 20 or 30 images you need to refine that way, it quickly becomes a huge time waste compared to the ease of an "add x steps" button. 25 is just usually too many unless you intentionally under-bake them the first time [e.g. initially only doing 15 or 20 steps on dpm2]... But that winds up wasting even more time if you know approximately how many steps you'd need the first time around.

**Comment by xeruf** [comment]
TLDR queue persistence and batch modifications, plus some wishes?

**Comment by spiralofhope** [comment]
It might be a good idea to break this request into multiple smaller issues.

In this case I'm highly interested in queue crash recovery.  It's horrifying to think my queue could be wiped.

**Comment by cmdr2** [comment]
A bunch of these have been fixed (using the "Use as input" button), and queue recovery is possible using the History plugin. And queue saving is being planned for a future version of the UI. Closing this for now. Thanks
